Hide read-only items in the Archive

JesseSDevaney opened the pull request:
**Problem:** Read-only items - items belonging to read-only collections - are appearing in users' archives. This causes frustration because they are shown items they can perform no actions against. Additionally, showing the items (as is) provide no context/value to the user.

**Context:**
- https://github.com/metabase/metabase/issues/24018

---
**Desired Behavior**/**Acceptance Criteria**

_Behaviors_
- [ ] Only actionable items - curate-able items - are shown in the archive
  - [ ] Read-only items are hidden
- [ ] Archive has no other altered behaviors

_Tests_
- [ ] Test added to verify curate-able items are **shown** and read-only items are **not shown** to users'
- [ ] Test added to verify admins can see all items

---
**How to Verify**

1. Log in as admin 
3. Create read-only collection > add question to read-only collection > archive the question
4. Create curate-able collection > add question to curate-able collection > archive the question
5. View the archive > see that both items appear and you can perform actions against both
  ![image](https://github.com/metabase/metabase/assets/22608765/4032c0c6-b04c-47c7-b82f-f2e7d8ea49d2)
6. Setup the archive again with a read-only item and curate-able item
7. Log in as read-only user
8. View the archive > see that only the curate-able item appears and that you are able to unarchive it
  ![image](https://github.com/metabase/metabase/assets/22608765/87364257-ca38-4970-b8c2-ee483f5b7eba)

JesseSDevaney committed:
convert to functional component

JesseSDevaney committed:
refactor use-list-select hook

- toggleAll(items) felt unintuitive as some items could be selected while simultaneously unselecting others
  - switched to selectOnlyTheseItems(items) to be more explicit on what the function is doing
- updated consumers of useListSelect to use the new API

JesseSDevaney committed:
convert HoC ListSelect to hook useListSelect

JesseSDevaney committed:
convert redux connects to useSelectors

JesseSDevaney committed:
refactor ArchiveApp and improve types

JesseSDevaney committed:
hide read only items in the archive

JesseSDevaney committed:
verify desired behavior

JesseSDevaney committed:
refactor beforeEach

JesseSDevaney committed:
remove unused eslint disable line

JesseSDevaney committed:
remove unused eslint disable rule

JesseSDevaney commented on e2e/test/scenarios/collections/archive.cy.spec.js:
```diff
@@ -15,6 +19,42 @@ describe("scenarios > collections > archive", () => {
+
+    createAndArchiveDashboard({
+      name: CURATEABLE_NAME,
+      collection_id: 6,
```
Is there a constant/variable with the readonly user's `collection-id`? I would prefer not to hard-code it.

JesseSDevaney committed:
Merge branch 'master' into fix-24018/hide-read-only-items-in-the-archive

JesseSDevaney committed:
refactor item can_write accessor logic

adam-james-v committed:
Add :can_write key into the search data response. uses `mi/can-write?`

adam-james-v commented on src/metabase/api/search.clj:
```diff
@@ -456,6 +456,7 @@
                             (map t2.realize/realize)
                             (map to-toucan-instance)
                             (filter check-permissions-for-model)
+                            (map #(assoc % :can_write (mi/can-write? %)))
```
@JesseSDevaney, adding this should help, I think. Now, running `api/search?archived=true`
Gives the normal response, but inside the :data key a result will look something like this:

```clojure
{:description nil,
 :archived true, ;; <----- even shows for archived true
 :collection_position nil,
 :table_id nil,
 :bookmark false,
 :can_write true, ;; <---- can_write for the result
 :database_id nil,
 :pk_ref nil,
 :name "Audit custom reports",
 :table_schema nil,
 :collection_authority_level nil,
 :updated_at nil,
 :moderated_status nil,
 :model_index_id nil,
 :model_id nil,
 :model_name nil,
 :id 113,
 :table_description nil,
 :dashboardcard_count nil,
 :context nil,
 :initial_sync_status nil,
 :table_name nil,
 :collection {:id 113, :name "Audit custom reports", :authority_level nil},
 :model "collection"}
```

JesseSDevaney replied:
Thank you so much for this, Adam! It works perfectly :heart:

JesseSDevaney committed:
Merge branch 'master' into fix-24018/hide-read-only-items-in-the-archive

JesseSDevaney committed:
update archive app for archived search changes

JesseSDevaney committed:
refactor filtering of list

JesseSDevaney committed:
replace hard-coded collection ids in spec

escherize commented:
Adding the `can_write?` information to search results makes sense to me. 👍

escherize reviewed (approved):
Approving the backend change.

iethree commented on e2e/support/cypress_sample_instance_data.js:
```diff
@@ -35,6 +35,11 @@ export const NORMAL_PERSONAL_COLLECTION_ID = _.findWhere(
   { name: "Robert Tableton's Personal Collection" },
 ).id;
 
+export const READ_ONLY_PERSONAL_COLLECTION_ID = _.findWhere(
+  SAMPLE_INSTANCE_DATA.collections,
+  { name: "Read Only Tableton's Personal Collection" },
+).id;
```
You... didn't hard code any IDs ❤️  🥹

iethree commented on e2e/support/helpers/e2e-dashboard-helpers.js:
```diff
@@ -205,6 +205,15 @@ export const dashboardHeader = () => {
   return cy.findByTestId("dashboard-header");
 };
 
+export const createAndArchiveDashboard = ({ name, collection_id }) => {
+  cy.request("POST", "/api/dashboard/", {
+    name: name,
+    collection_id: collection_id,
+  }).then(({ body: { id: dashboardId } }) => {
+    cy.request("PUT", `/api/dashboard/${dashboardId}`, { archived: true });
+  });
+};
```
nit, if you're only using this helper in one spec, I'd just keep it in that spec file. I doubt we'll be using this helper widely.

iethree commented on frontend/src/metabase/archive/containers/ArchiveApp.tsx:
```diff
@@ -53,16 +53,27 @@ export function ArchiveApp() {
+      return [];
+    }
+
+    if (isAdmin) {
+      return data;
+    }
```
I'm on the fence about this, but I think we should delete this check. With Auditv2 we've had to remove and change a lot of logic that assumes that admins can edit anything since that's no longer true. In its current form, Audit collections shouldn't be archive-able, but I'd like to move to a world where we just rely on `can_write` instead of assuming that admins can edit anything.

iethree reviewed (approved):
This looks good to me.

Question though: Did you discuss adding a query param to the search endpoint to only fetch writable items? It seems strange to fetch a bunch of irrelevant data on the frontend and do a bunch of filtering to hide it.

JesseSDevaney committed:
move specific helper outside of generic helpers file

JesseSDevaney committed:
remove admin gets unfiltered data

adam-james-v committed:
BE sends only can_write true results when search with archived?=true

This change modifies the `check-permissions-for-model` method to check for write access if the SearchContext has
archived true.

This means that the only results sent to the frontend are those results that the user can take actions on, namely,
unarchiving them.

adam-james-v committed:
Merge branch 'master' into fix-24018/hide-read-only-items-in-the-archive

adam-james-v committed:
reomve unused require

adam-james-v committed:
Merge branch 'master' into fix-24018/hide-read-only-items-in-the-archive

JesseSDevaney committed:
update FE to account for BE changes
